File: ruby/run-length-encoding/run_length_encoding.rb
class RunLengthEncoding
  def self.encode(input)
    input
      .chars
      .slice_when { |i, j| i != j }
      .map { |i| i.size > 1 ? [i.size, i[0]] : i[0] }
      .join
  end

  def self.decode(input)
    input.gsub(/[0-9]+[A-Za-z ]/) { |match| match[-1] * match.to_i }
  end
end

module BookKeeping
  VERSION = 3
end

File: JavaScript/run-length-encoding/run-length-encoding.js
export function encode(input) {
  let encoding = [];
  let previous = input[0];
    let count = 1; 
    for (let i = 1; i <= input.length; i++) {
        if (input[i] !== previous) {
            encoding.push([count, previous]);
            count = 1;
            previous = input[i];
        }
        else {
            count ++;
        }
    }
    return encoding.map(([reps, char]) => reps == 1? char : reps + char).join('');
}
export function decode(input) {
    const countedChars = /(\d+)(\w|\s)/g;
    return input.replace(countedChars, (_, repeats, char) => new Array( + repeats + 1).join(char));
}